--- VirtualBox-6.1.0_BETA2/include/iprt/cdefs.h
+++ VirtualBox-6.1.0_BETA2/include/iprt/cdefs.h
@@ -1170,7 +1170,7 @@
  * Tell the compiler that we're falling through to the next case in a switch.
  * @sa RT_FALL_THRU  */
 #if RT_GNUC_PREREQ(7, 0)
-# define RT_FALL_THROUGH()      __attribute__((fallthrough))
+# define RT_FALL_THROUGH()      __attribute__((__fallthrough__))
 #else
 # define RT_FALL_THROUGH()      (void)0
 #endif
