--- VirtualBox-6.1.0_BETA2/src/VBox/Runtime/common/fs/ntfsvfs.cpp	2019-10-23 10:32:51.000000000 -0400
+++ VirtualBox-6.1.0_BETA2/src/VBox/Runtime/common/fs/ntfsvfs.cpp	2019-11-07 08:03:01.716327464 -0500
@@ -1052,7 +1052,7 @@
                     case 4: cClustersInRun |= (uint32_t)pbPairs[offPairs + 3] << 24; RT_FALL_THRU();
                     case 3: cClustersInRun |= (uint32_t)pbPairs[offPairs + 2] << 16; RT_FALL_THRU();
                     case 2: cClustersInRun |= (uint16_t)pbPairs[offPairs + 1] <<  8; RT_FALL_THRU();
-                    case 1: cClustersInRun |= (uint16_t)pbPairs[offPairs + 0] <<  0; RT_FALL_THRU();
+                    case 1: cClustersInRun |= (uint16_t)pbPairs[offPairs + 0] <<  0;
                 }
                 offPairs += cbRunField;
                 AssertBreakStmt(cClustersInRun <= cMaxClustersInRun,
