--- a/src/VBox/HostDrivers/VBoxNetFlt/linux/VBoxNetFlt-linux.c	2021-03-21 16:51:36.286139986 -0400
+++ b/src/VBox/HostDrivers/VBoxNetFlt/linux/VBoxNetFlt-linux.c	2021-03-21 16:55:13.771622954 -0400
@@ -48,6 +48,9 @@
 #if RTLNX_VER_MIN(4,5,0)
 #include <uapi/linux/pkt_cls.h>
 #endif
+#if RTLNX_VER_MIN(5,11,0)
+#include <linux/ethtool.h>
+#endif
 #include <net/ipv6.h>
 #include <net/if_inet6.h>
 #include <net/addrconf.h>
